core/src/ser.rs

```diff
@@ -223,14 +239,35 @@ pub trait ByteFormat {
         W: Write;
 
     /// Read a sequence of exactly `limit` objects from the reader.
-    fn read_seq_from<R>(reader: &mut R, limit: usize) -> Result<Vec<Self>, Self::Error>
+    fn read_seq_from<R>(reader: &mut R, mode: ReadSequenceMode) -> Result<Vec<Self>, Self::Error>
     where
         R: Read,
         Self: std::marker::Sized,
     {
         let mut v = vec![];
-        for _ in 0..limit {
-            v.push(Self::read_from(reader)?);
+        match mode {
+            ReadSequenceMode::Exactly(number) => {
```

Comment:
Maybe worth using `Vec::with_capacity` in this block

Reply:
that sets the cap on vec size without reallocating. We need to set an exact allocated size